File: java/src/main/java/com/laamella/code_state_machine/Conditions.java
package com.laamella.code_state_machine;

import com.laamella.code_state_machine.util.Chain;

/**
 * A simple wrapper around a list of conditions.
 */
public class Conditions<E> extends Chain<Condition<E>> {
    // This method exists only to suppress warnings about varargs.
    public Conditions() {
        super();
    }

    @SafeVarargs
    public Conditions(Condition<E>... conditions) {
        super(conditions);
    }

    public void handleEvent(E event) {
        for (var condition : getItems()) {
            condition.handleEvent(event);
        }
    }

    /**
     * @return true if all conditions are met, else false.
     */
    public boolean isMet() {
        for (var condition : getItems()) {
            if (!condition.isMet()) {
                return false;
            }
        }
        return true;
    }

    public void reset() {
        for (var condition : getItems()) {
            condition.reset();
        }
    }
}

File: kotlin/src/main/java/com/laamella/kode_state_machine/Conditions.kt
package com.laamella.kode_state_machine

import com.laamella.kode_state_machine.util.Chain

/**
 * A simple wrapper around a list of conditions.
 */
class Conditions<E> : Chain<Condition<E>> {
    // This method exists only to suppress warnings about varargs.
    constructor() : super()

    @SafeVarargs
    constructor(vararg conditions: Condition<E>) : super(*conditions)

    fun handleEvent(event: E) {
        for (condition in items) {
            condition.handleEvent(event)
        }
    }

    val isMet: Boolean
        /**
         * @return true if all conditions are met, else false.
         */
        get() {
            for (condition in items) {
                if (!condition.isMet) {
                    return false
                }
            }
            return true
        }

    fun reset() {
        for (condition in items) {
            condition.reset()
        }
    }
}
